---
title: "V(D)J analysis"
date: '`r format(Sys.Date(), "%B %d %Y")`'
output:
  rmarkdown::html_vignette:
    toc_float: true
    toc_depth: 2
    vignette: >
      %\VignetteIndexEntry{V(D)J analysis}
      %\VignetteEncoding{UTF-8}
      %\VignetteEngine{knitr::rmarkdown}
---

```{r "setup", include = FALSE}
# Chunk opts
knitr::opts_chunk$set(
  collapse  = TRUE,
  comment   = "#>",
  warning   = FALSE,
  message   = FALSE
)

# Packages
library(stringr)
library(tibble)
library(purrr)
library(dplyr)
library(ggplot2)
library(cowplot)
library(Seurat)
library(here)
library(djvdj)

# Color palette
vdj_colors <- c(
  "#E69F00", "#56B4E9",
  "#009E73", "#F0E442",
  "#d7301f", "#0072B2",
  "#D55E00", "#6A51A3",
  "#CC79A7", "#999999",
  "#875C04"
)
```

<br>

For this vignette we are using AVID-seq data. This is a method developed by the Hesselberth lab which involves staining cells with DNA-tagged antigen. The DNA tag is similar to the tags present on CITE-seq antibodies and can be detected using the 10x Genomics 5â€™ immune profiling kit. For this experiment we mixed splenocytes from BL6 and MD4 mice and stained with a HEL-DNA conjugate. MD4 B cells are monoclonal and specifically bind HEL.

```{r "rna_umap", fig.width = 12.5, fig.height = 5, echo = FALSE}
load(here("data/avid/so_avid.rda"))

# Rename cell types
so_avid <- so_avid %>%
  mutate_meta(
    .fun = mutate,
    type = if_else(!type %in% c("B cell", "T cell"), "other", type)
  )

# Plot colors
m_cols <- c(
  BL6 = "#E69F00",
  MD4 = "#56B4E9"
)

t_cols <- c(
  "B cell" = "#56B4E9",
  "T cell" = "#009E73",
  other    = "#F0E442"
)

# Create mouse UMAP
mouse_umap <- so_avid %>%
  plot_features(
    feature     = "mouse",
    plot_colors = m_cols
  ) +
  ggtitle("Mouse") +
  theme_nothing() +
  theme(
    plot.title      = element_text(),
    legend.position = "right",
    legend.title    = element_blank()
  )
  
# Create cell type UMAP
hel_umap <- so_avid %>%
  plot_features(
    feature     = "type",
    plot_colors = t_cols,
    feat_lvls   = names(t_cols)
  ) +
  ggtitle("Cell type") +
  theme_nothing() +
  theme(
    plot.title      = element_text(),
    legend.position = "right",
    legend.title    = element_blank()
  )

# Final figure
plot_grid(
  mouse_umap, hel_umap,
  rel_widths = c(1, 0.9)
)
```

<br>

Signal from the HEL-DNA conjugate is shown below. Most signal is localized within MD4 B cells.

```{r "hel_umap", fig.width = 6, fig.height = 5, echo = FALSE}
so_avid %>%
  plot_features(
    feature     = "adt_HEL",
    plot_colors = c("white", vdj_colors[5]),
    pt_size     = 1,
    pt_outline  = 0.2
  ) +
  ggtitle("HEL signal") +
  guides(color = guide_colorbar(barwidth = unit(6, "pt"))) +
  theme_nothing() +
  theme(
    plot.title      = element_text(),
    legend.position = "right",
    legend.title    = element_blank()
  )
```

<br>

## Import VDJ data

`import_vdj()` takes the output files from [Cell Ranger](https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/vdj#header) and adds clonotype information to the meta.data for an existing [Seurat](https://satijalab.org/seurat/) object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a ';' (or a character specified by `sep`). For cells that do not have any VDJ sequencing data, `NA`s will be added to the meta.data.

If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector. 

```{r "import_vdj"}
# Create vector of paths pointing to cellranger output
paths <- here("data/avid/bcr")

so_avid <- import_vdj(
  sobj_in        = so_avid,                 # Seurat object
  vdj_dir        = paths,                   # cellranger directories
  filter_contigs = TRUE                     # Only include productive chains
)
```

<br>

Take a look at the meta.data to see the V(D)J data added to the object. 

```{r}
vdj_cols <- c(
  "clonotype_id", "cdr3",
  "chains", "v_gene", 
  "j_gene", "reads",
  "umis"
)

so_avid@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

## Quality Control

### Read Support

The read support for each chain can be visualized with the `plot_reads()` function. This will create plots summarizing the number of UMIs and total reads that were obtained for each chain.

```{r "read_support", fig.width = 5, fig.height = 3}
plot_reads(
  sobj_in     = so_avid,                    # Seurat object
  chain_col   = "chains",                   # Column containing chains for each cell
  plot_colors = vdj_colors
) +
  guides(fill = FALSE, color = FALSE)
```

<br>

### Paired Chains

The djvdj package provides several functions that can be used to manipulate the object meta.data. This includes a function to modify meta.data columns (`mutate_vdj()`) and a function to filter cells from the object (`filter_vdj()`). Since cells can have multiple values present in each V(D)J column, when using these functions each string of values for a cell should be thought of as a vector.

The V(D)J data imported from Cell Ranger may include clonotypes that do not have paired heavy and light chains. Using the `mutate_vdj()` function, we can add a new column (`uniq_chains`) to the meta.data to shows the unique chains detected in each cell.

```{r}
so_avid <- mutate_vdj(
  sobj_in     = so_avid,                               # Seurat object
  uniq_chains = str_c(unique(chains), collapse = "_")  # New column
)

# Take a look at the meta.data
so_avid@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id), n_chains > 2) %>%
  select(all_of(vdj_cols), uniq_chains)
```

For row 2 shown above, this is equivalent to:

```{r}
vec <- c("IGH", "IGH", "IGK", "IGK")
vec <- unique(vec)

str_c(vec, collapse = "_")
```

<br>

We can then overlay the new labels present in the `uniq_chains` column on a UMAP projection. djvdj also provides a simple wrapper function (`plot_cell_count()`) that will summarize the fraction of cells that belong to each group.

```{r "chains_umap", fig.width = 6.5, fig.height = 4.2}
chain_cols <- vdj_colors
names(chain_cols) <- unique(so_avid$uniq_chains)

# Chains UMAP
chain_umap <- so_avid %>%
  DimPlot(
    group.by = "uniq_chains",
    cols     = chain_cols,
    na.value = "grey90"
  ) +
  theme_nothing() +
  theme(legend.position = "top")

# Chains bar graphs
chain_bars <- so_avid %>%
  plot_cell_count(
    x           = "type",                   # Cell labels to plot
    fill_col    = "uniq_chains",            # Cell labels to color each bar
    plot_colors = chain_cols,
    n_label     = FALSE
  ) +
  theme(
    plot.margin     = unit(c(120, 10, 10, 10), "pt"),
    legend.position = "none"
  )

# Final figure
plot_grid(
  chain_umap, chain_bars,
  rel_widths = c(1, 0.55),
  nrow       = 1
)
```

<br>

The `filter_vdj()` function can be used to remove V(D)J data for cells that lack paired chains. Setting `filter_cells` to `FALSE` (the default) will keep all cells in the object, but the V(D)J information will be removed.

```{r}
so_filt <- filter_vdj(
  sobj_in      = so_avid,                                                # Seurat object
  filt         = "IGH" %in% chains && any(chains %in% c("IGK", "IGL")),  # Condition for filtering
  filter_cells = FALSE                                                   # Should cells be removed
)
```

<br>

## Clonotype Abundance

To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the `calc_abundance()` function. These calculations can be performed on a per-cluster or per-sample basis by also providing a meta.data column containing cell labels.

```{r}
so_avid <- calc_abundance(
  sobj_in     = so_avid,                    # Seurat object
  cluster_col = "mouse"                     # meta.data column for grouping
)
```

<br>

For each 'calc' function provided by djvdj, there is a matching 'plot' function that will generate a summary plot. The `plot_abundance()` function will plot clonotypes ranked by abundance. As expected we see that most MD4 B cells share the same clonotype ID.

```{r "abund_plots", fig.width = 8, fig.height = 3.5}
plot_abundance(
  sobj_in      = so_avid,                   # Seurat object
  cluster_col  = "mouse",                   # meta.data column for grouping
  label_col    = "clonotype_id",            # meta.data column containing labels
  n_clonotypes = 12,                        # Number of top clonotypes
  plot_colors  = m_cols
) +
  theme(legend.position = "none")
```

<br>

## Repertoire Diversity

The function `calc_diversity()` will calculate repertoire diversity on number of cells that share each clonotype. Using the `cluster_col` argument, any meta.data column containing cell labels can be used for calculations. `calc_diversity()` uses the R package [abdiv](https://github.com/kylebittinger/abdiv) for performing diversity calculations and any [abdiv](https://github.com/kylebittinger/abdiv) diversity function can be specified using the `method` argument. It is important to read the [abdiv](https://github.com/kylebittinger/abdiv) documentation to ensure the selected function is appropriate for your analysis. 

Possible methods for calculating diversity include:

```{r, comment = "", echo = F}
abdiv::alpha_diversities
```
<br>

In this example we are calculating the Shannon diversity for each sample in the orig.ident meta.data column.

```{r}
so_avid <- calc_diversity(
  sobj_in     = so_avid,                    # Seurat object
  cluster_col = "mouse",                    # meta.data column for grouping
  method      = abdiv::shannon              # abdiv method to use
)
```

<br>

The `plot_diversity()` function will create plots summarizing repertoire diversity for each sample. A named list of functions to run can also be passed to plot multiple metrics. Two different metrics are shown in the example below. As expected, BL6 B cells have a very diverse repertoire, while MD4 cells show low diversity values. 

```{r "div_plots", fig.width = 3.5, fig.height = 3}
# Metrics to plot
fns <- list(
  "simpson"  = abdiv::simpson,
  "mcintosh" = abdiv::mcintosh_d
)

plot_diversity(
  sobj_in     = so_avid,                    # Seurat object
  cluster_col = "mouse",                    # meta.data column for grouping
  method      = fns,                        # abdiv method to use
  plot_colors = vdj_colors
) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

<br>

## Repertoire Overlap

To compare repertoires for different samples or clusters, `calc_similarity()` can calculate a variety of different similarity metrics. The `cluster_col` should be used to specify the meta.data column containing cell labels for comparison. Like `calc_diversity()`, an [abdiv](https://github.com/kylebittinger/abdiv) function can be specified with the `method` argument. It is important to read the [abdiv](https://github.com/kylebittinger/abdiv) documentation to ensure the selected function is appropriate for your analysis.

Possible methods for calculating repertoire similarity include:

```{r, comment = "", echo = F}
abdiv::beta_diversities
```

<br>

By default `calc_similarity()` will add a new meta.data column for each comparison. In this example we are calculating the jaccard dissimilarity index for all combinations of clusters present in the `seurat_clusters` column.

```{r}
so_avid <- calc_similarity(
  sobj_in     = so_avid,                    # Seurat object
  cluster_col = "seurat_clusters",          # meta.data column for grouping
  method      = abdiv::jaccard              # abdiv method to use
)
```

<br>

A heatmap summarizing the results can be generated using the `plot_similarity()` function. Values closer to 1 indicate the repertoires are very different.

```{r "sim_plots", fig.width = 5, fig.height = 4}
plot_similarity(
  sobj_in     = so_avid,                    # Seurat object
  cluster_col = "seurat_clusters",          # meta.data column for grouping
  method      = abdiv::jaccard,             # Method to use
  plot_colors = "#009E73",
  size        = 0.2,                        # Additional ggplot options
  color       = "white"                     # Additional ggplot options
) +
  theme(
    legend.text = element_text(size = 8),
    axis.text.x = element_text(angle = 0)
  )
```

<br>

## Gene Usage

The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function `calc_usage()` can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the `chain` and `chain_col` arguments can be used to specify the meta.data column containing the chains for each cell. By default the results for all chains will be included.

In this example we are summarizing the usage of different V genes for the TRB chain 

```{r}
calc_usage(
  sobj_in     = so_avid,                    # Seurat object
  gene_cols   = "v_gene",                   # meta.data column containing genes
  cluster_col = "mouse",                    # meta.data column for grouping
  chain       = "IGH",                      # Chain to use for filtering genes
  chain_col   = "chains"                    # meta.data column containing chains
)
```

<br>

The function `plot_usage()` can be used to summarize these results. Using the `yaxis` argument, the percent or absolute count (frequency) can be used for plotting. The genes plotted can also be selected using the `plot_genes` argument, or the number of top genes (most frequent) to plot can be specified with `n_genes`.

```{r "gene_usage_1", fig.width = 11, fig.height = 3.5}
plot_usage(
  sobj_in     = so_avid,                    # Seurat object
  gene_cols   = "v_gene",                   # meta.data column(s) containing genes
  type        = "bar",                      # Type of plot
  chain       = "IGH",                      # Chain to use for filtering genes
  
  plot_genes  = NULL,                       # A list of genes to plot
  n_genes     = 50,                         # The number of top genes to plot
  plot_colors = "#0072B2"
)
```

<br>

By passing multiple columns to `gene_cols`, the frequency that different genes are used together can also be summarized.

```{r, eval = F}
calc_usage(
  sobj_in     = so_avid,                    # Seurat object
  gene_cols   = c("v_gene", "j_gene"),      # meta.data column(s) containing genes
  cluster_col = "mouse",                    # meta.data column for grouping
  chain       = "IGK"                       # Chain to use for filtering genes
)
```

<br>

When multiple gene columns are passed to `plot_usage()`, a list of plots will be returned, one for each cell label in the `cluster_col` column.

```{r "gene_usage_2", fig.width = 12, fig.height = 2.5}
ggs <- plot_usage(
  sobj_in     = so_avid,                    # Seurat object
  gene_cols   = c("v_gene", "j_gene"),      # meta.data column(s) containing genes
  cluster_col = "mouse",                    # meta.data column for grouping
  chain       = "IGK",                      # Chain to use for filtering genes
  plot_colors = "#6A51A3",
  n_genes     = 20
) %>%
  imap(~ .x + ggtitle(.y))

plot_grid(plotlist = ggs)
```

<br>

## CDR3 Usage

The `calc_usage()` and `plot_usage()` functions can also be used to calculate the fraction of cells that have a certain CDR3 sequence. To do this, the meta.data column containing the CDR3 sequences for each cell can be passed to the `gene_cols` argument. The plot can be divided into facets based on sample using simple ggplot2 commands.

```{r "cdr3_usage", fig.width = 8, fig.height = 3.5}
plot_usage(
  sobj_in     = so_avid,
  gene_cols   = "cdr3",                     # Column containing CDR3 sequences
  cluster_col = "mouse",
  
  type        = "bar",
  chain       = "IGK",
  plot_colors = vdj_colors,
  n_genes     = 12                          # Number of CDR3 sequences to plot
) +
  facet_wrap(~ mouse, nrow = 1, scales = "free_x") +
  theme(legend.position = "none")
```

